<a href="https://colab.research.google.com/github/Younes-Karimi/IST597-DataPrivacy/blob/main/DP_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [45]:
import pandas as pd
import numpy as np
from google.colab import drive
import matplotlib.pyplot as plt
drive.mount('/content/drive')
dir = '/content/drive/My Drive/data_privacy/Adult/'
pd.options.mode.chained_assignment = None  # default='warn'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


instead what i could do is generalize first and then do this method numerically, and then convery back to the category types

import data and rename - to _

In [3]:
df = pd.read_table(dir+'dataset_full.txt', delim_whitespace=True, names=('age', 'education-num', 'work-class','marital-status','occupation','race','sex','native-country'))
df = df.rename(columns={'education-num':'education_num','work-class':'work_class','marital-status':'marital_status','native-country':'native_country'})

In [4]:
df.columns

Index(['age', 'education_num', 'work_class', 'marital_status', 'occupation',
       'race', 'sex', 'native_country'],
      dtype='object')

i am choosing age, native country, and marital status as the quasi-identifiers  that we want to generalize and occupation as the sensitive column

age and education num are numerical data and the rest are categorical data

# Dataset Analysis

List the column names in the dataset. For each column, identify the type of information (string, integer, etc.) and present the basic descriptives (range of values, mean, frequencies, outliers or unique values, etc). Explain the utility of the data that you want to preserve after your anonymization process. For example, discuss what types of queries you want your anonymized data to support. Your anonymization approach should support your answer.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30162 entries, e1 to e30162
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   education_num   30162 non-null  int64 
 2   work_class      30162 non-null  object
 3   marital_status  30162 non-null  object
 4   occupation      30162 non-null  object
 5   race            30162 non-null  object
 6   sex             30162 non-null  object
 7   native_country  30162 non-null  object
dtypes: int64(2), object(6)
memory usage: 2.1+ MB


This dataset consists of 8 columns and 30162 entries as shown above. The first two columns, *age* and *education_num*, are integers while the rest of columns are objects and have categorical (string) values.

In [78]:
print('Min age: {}\nMax age: {}'.format(df['age'].values.min(), df['age'].values.max()))
print('Mean age: %.2f' % (df['age'].values.sum() / df['age'].values.size))
print('\nSorted unique values: \n', np.sort(df['age'].unique()))
print('\nAge frequencies:\n', df['age'].value_counts(sort=True))

Min age: 17
Max age: 90
Mean age: 38.44

Sorted unique values: 
 [17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40
 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64
 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 88 90]

Age frequencies:
 36    852
31    851
33    837
34    836
35    828
     ... 
82      7
83      5
88      3
85      3
86      1
Name: age, Length: 72, dtype: int64


In [76]:
print('Min education number: {} year(s)\nMax education number: {} year(s)'.format(df['education_num'].values.min(), df['education_num'].values.max()))
print('Mean education number: %.2f year(s)' % (df['education_num'].values.sum() / df['education_num'].values.size))
print('\nEducation number frequencies:\n', df['education_num'].value_counts(sort=True))

Min education number: 1 year(s)
Max education number: 16 year(s)
Mean education number: 10.12 year(s)

Education number frequencies:
 9     9840
10    6678
13    5044
14    1627
11    1307
7     1048
12    1008
6      820
4      557
15     542
5      455
8      377
16     375
3      288
2      151
1       45
Name: education_num, dtype: int64


In [80]:
print('Work class frequencies:\n\n', df['work_class'].value_counts(sort=True))

Work class frequencies:

 Private             22286
Self-emp-not-inc     2499
Local-gov            2067
State-gov            1279
Self-emp-inc         1074
Federal-gov           943
Without-pay            14
Name: work_class, dtype: int64


In [81]:
print('Marital status frequencies:\n\n', df['marital_status'].value_counts(sort=True))

Marital status frequencies:

 Married-civ-spouse       14065
Never-married             9726
Divorced                  4214
Separated                  939
Widowed                    827
Married-spouse-absent      370
Married-AF-spouse           21
Name: marital_status, dtype: int64


In [83]:
print('Occupation frequencies:\n\n', df['occupation'].value_counts(sort=True))

Occupation frequencies:

 Prof-specialty       4038
Craft-repair         4030
Exec-managerial      3992
Adm-clerical         3721
Sales                3584
Other-service        3212
Machine-op-inspct    1966
Transport-moving     1572
Handlers-cleaners    1350
Farming-fishing       989
Tech-support          912
Protective-serv       644
Priv-house-serv       143
Armed-Forces            9
Name: occupation, dtype: int64


In [86]:
print('Race frequencies:\n\n', df['race'].value_counts(sort=True))

Race frequencies:

 White                 25933
Black                  2817
Asian-Pac-Islander      895
Amer-Indian-Eskimo      286
Other                   231
Name: race, dtype: int64


In [85]:
print('Sex frequencies:\n\n', df['sex'].value_counts(sort=True))

Sex frequencies:

 Male      20380
Female     9782
Name: sex, dtype: int64


In [84]:
print('Native country frequencies:\n\n', df['native_country'].value_counts(sort=True))

Native country frequencies:

 United-States                 27504
Mexico                          610
Philippines                     188
Germany                         128
Puerto-Rico                     109
Canada                          107
India                           100
El-Salvador                     100
Cuba                             92
England                          86
Jamaica                          80
South-Africa                     71
China                            68
Italy                            68
Dominican-Republic               67
Vietnam                          64
Guatemala                        63
Japan                            59
Columbia                         56
Poland                           56
Taiwan                           42
Haiti                            42
Iran                             42
Portugal                         34
Nicaragua                        33
Peru                             30
Greece                           2

# Identification of Sensitive Information

can convert the categorical data to 'category' in pandas for easier handling

In [ ]:
df['work_class'] = df['work_class'].astype('category')
work_class_dict = dict(enumerate(df['work_class'].cat.categories))
df['work_class'] = df['work_class'].astype('category').cat.codes

df['marital_status'] = df['marital_status'].astype('category')
marital_status_dict = dict(enumerate(df['marital_status'].cat.categories))
df['marital_status'] = df['marital_status'].astype('category').cat.codes

df['occupation'] = df['occupation'].astype('category')
occupation_dict = dict(enumerate(df['occupation'].cat.categories))
df['occupation'] = df['occupation'].astype('category').cat.codes

df['race'] = df['race'].astype('category')
race_dict = dict(enumerate(df['race'].cat.categories))
df['race'] = df['race'].astype('category').cat.codes

df['sex'] = df['sex'].astype('category')
sex_dict = dict(enumerate(df['sex'].cat.categories))
df['sex'] = df['sex'].astype('category').cat.codes

df['native_country'] = df['native_country'].astype('category')
native_country_dict = dict(enumerate(df['native_country'].cat.categories))
df['native_country'] = df['native_country'].astype('category').cat.codes

going off of the mondrian algorithm, which relies on paritioning the data into k-anonymous groups following a greedy algorithm

In [ ]:
df.index

Index(['e1', 'e2', 'e3', 'e4', 'e5', 'e6', 'e7', 'e8', 'e9', 'e10',
       ...
       'e30153', 'e30154', 'e30155', 'e30156', 'e30157', 'e30158', 'e30159',
       'e30160', 'e30161', 'e30162'],
      dtype='object', length=30162)

In [ ]:
def attribute_domain(df, column, part):
  domain = df[column][part].max() - df[column][part].min()
  return domain

In [ ]:
def is_k_anon(df, part, sensitive_attribute, k=3):
  if len(part) < k:
    return False
  
  else:
    return True

In [ ]:
def partitioning(df, feature_columns, sensitive_attribute, is_k_anon):

    final_parts = []
    parts = [df.index]
    while parts:
        part = parts.pop(0)
        domains = {}
        for column in df:
          dom = attribute_domain(df, column, part)
          domains[column] = dom
        for column, dom in sorted(domains.items(), reverse=True):
            #split the column by the media
            df_part = df[column][part]
            median = df_part.median()
            lhs = df_part.index[df_part < median]
            rhs = df_part.index[df_part>= median]
            #check if the partitions satify k anon
            if (is_k_anon(df, lhs, sensitive_attribute) and is_k_anon(df, rhs, sensitive_attribute)):
              parts.extend((lhs, rhs))
            else:
              continue

            break
        else:
            final_parts.append(part)
    return final_parts

In [ ]:
#apply
QI_columns = ['native_country', 'marital_status','age','education_num']
sensitive_attribute = 'work_class'
final_partitions = partitioning(df, QI_columns, sensitive_attribute, is_k_anon)



In [ ]:
len(final_partitions)

4783

try to apply to data now

In [ ]:
def anonymize_data( df, parts, QI_columns, sensitive_attribute, max_partitions=None):
  aggregates ={}
  for column in QI_columns:
    aggregates[column] = [df[column].mean()]
  rows = []
  #print(parts)
  for i, part in enumerate(parts):
    print(part)
    if i % 100 == 1:
      print("Finished {} partitions...".format(i))
    if max_partitions is not None and i > max_partitions:
      break
    grouped_columns = df.loc[part].groupby(QI_columns, as_index=False).mean() 
    #print(df.loc[part].groupby(sensitive_attribute)['work_class'].head())
    sensitive_counts = df.loc[part].groupby(sensitive_attribute).agg({sensitive_attribute : 'count'})
    values = grouped_columns.iloc[0].to_dict()
    for sensitive_value, count in sensitive_counts[sensitive_attribute].items():
      if count == 0:
        continue
      values.update({sensitive_attribute : sensitive_value,'count' : count,})
      rows.append(values.copy())
  return pd.DataFrame(rows)


In [ ]:
dfn = anonymize_data(df, final_partitions, QI_columns, sensitive_attribute)

Index(['e2026', 'e18685', 'e19193'], dtype='object')
Index(['e3748', 'e10539', 'e13450', 'e18764'], dtype='object')
Finished 1 partitions...
Index(['e16432', 'e22759', 'e26151'], dtype='object')
Index(['e4871', 'e8517', 'e9797', 'e12041', 'e12658'], dtype='object')
Index(['e14496', 'e19429', 'e21806', 'e25180'], dtype='object')
Index(['e2445', 'e3675', 'e19493', 'e24385', 'e25982'], dtype='object')
Index(['e8963', 'e25868', 'e29936'], dtype='object')
Index(['e8053', 'e9216', 'e9922', 'e16971', 'e28923'], dtype='object')
Index(['e3362', 'e16867', 'e18123', 'e23907'], dtype='object')
Index(['e1681', 'e2192', 'e8043', 'e10776', 'e14915', 'e22952', 'e28786'], dtype='object')
Index(['e7153', 'e9330', 'e17294', 'e22547', 'e24198'], dtype='object')
Index(['e3478', 'e16633', 'e16647', 'e19523', 'e22273'], dtype='object')
Index(['e5485', 'e8499', 'e8714'], dtype='object')
Index(['e8767', 'e19578', 'e22993'], dtype='object')
Index(['e1237', 'e2207', 'e7642', 'e21022'], dtype='object')
Index(['e2

KeyboardInterrupt: ignored

In [ ]:
df.loc[['e532', 'e4823', 'e7492', 'e11207', 'e22125']]

age  education_num  work_class  ...  race  sex  native_country
e532     36             10           1  ...     4    0              38
e4823    41             10           1  ...     4    0              38
e7492    32              9           1  ...     4    0              38
e11207   45              9           1  ...     4    0              38
e22125   53              9           1  ...     4    0              38

[5 rows x 8 columns]

In [ ]:
dfn.sort_values(QI_columns+[sensitive_attribute])

native_country  marital_status   age  ...  race  sex  count
687              0.0             2.0  27.0  ...   1.0  1.0      3
874              0.0             2.0  32.0  ...   1.0  0.0      4
1135             0.0             2.0  37.0  ...   1.0  1.0      4
1136             0.0             2.0  37.0  ...   1.0  1.0      4
118              0.0             2.0  40.0  ...   1.0  1.0      6
...              ...             ...   ...  ...   ...  ...    ...
2643            38.0             6.0  77.0  ...   4.0  0.0      3
2644            38.0             6.0  77.0  ...   4.0  0.0      1
4966            38.0             6.0  77.0  ...   4.0  0.0      4
4967            38.0             6.0  77.0  ...   4.0  0.0      1
1787            39.0             4.0  23.0  ...   1.0  1.0      3

[7007 rows x 9 columns]

In [ ]:
dfn['age'].unique()

array([72., 46., 39., 31., 30., 53., 29., 36., 26., 22., 37., 35., 38.,
       42., 17., 64., 45., 62., 43., 33., 21., 40., 25., 54., 56., 63.,
       41., 49., 47., 24., 34., 18., 51., 27., 32., 44., 28., 23., 52.,
       19., 50., 65., 55., 20., 59., 57., 60., 48., 73., 66., 58., 68.,
       67., 61., 70., 77., 69., 76., 71., 85., 80., 78., 74., 75., 79.,
       90.])

In [ ]:
aggregates ={}
for column in QI_columns:
  aggregates[column] = df[column].mean()

In [ ]:
df.loc[['e2026', 'e18685', 'e19193']].agg({'marital_status': 2.5801339433724553, 'native_country': 36.382567469000726})

TypeError: ignored

In [ ]:
aggregates

{'marital_status': 2.5801339433724553, 'native_country': 36.382567469000726}

In [ ]:
df.loc[list(['e2026', 'e18685', 'e19193'])].agg(aggregates, squeeze=False)

TypeError: ignored

In [ ]:
parts = [['e2026', 'e18685', 'e19193']]

In [ ]:
df.groupby(['A', 'B'], as_index=False).sum()

In [ ]:
max_partitions = None
aggregates ={}
for column in QI_columns:
  aggregates[column] = [df[column].mean()]
rows = []
  #print(parts)
for i, part in enumerate(parts):
  if i % 100 == 1:
    print("Finished {} partitions...".format(i))
  if max_partitions is not None and i > max_partitions:
    break
  print(df.loc[part])
  grouped_columns = df.loc[part].groupby(QI_columns, as_index=False).mean()  #.agg(aggregates, squeeze=False)
  print(grouped_columns)
  sensitive_counts = df.loc[part].groupby(sensitive_attribute).agg({sensitive_attribute : 'count'})
  values = grouped_columns.iloc[0].to_dict()
  for sensitive_value, count in sensitive_counts[sensitive_attribute].items():
    if count == 0:
      continue
    values.update({sensitive_attribute : sensitive_value,'count' : count,})
    rows.append(values.copy())

        age  education_num  work_class  ...  race  sex  native_country
e2026    42             10           0  ...     4    0              32
e18685   72              7           0  ...     4    0               1
e19193   61              9           0  ...     4    0              32

[3 rows x 8 columns]
   age  marital_status  native_country  ...  occupation  race  sex
0   42               0              32  ...           0     4    0
1   61               4              32  ...           0     4    0
2   72               0               1  ...           0     4    0

[3 rows x 8 columns]


In [ ]:
aggregates

{'age': [38.437901995888865],
 'marital_status': [2.5801339433724553],
 'native_country': [36.382567469000726]}

In [ ]:
df.loc[part].agg({'native_country': [36.382567469000726]}, squeeze=False)

ValueError: ignored